## Imports and Loading

In [105]:
import numpy as np
import pandas as pd

In [106]:
df = pd.read_csv("../../datasets/messy_employee_dataset/Messy_Employee_dataset.csv")

print(f"Loaded dataset: {df.shape[0]} rows, {df.shape[1]} columns")

Loaded dataset: 1020 rows, 12 columns


In [107]:
df.head()

,Employee_ID,First_Name,Last_Name,Age,Department_Region,Status,Join_Date,Salary,Email,Phone,Performance_Score,Remote_Work
0,EMP1000,Bob,Davis,25.0,DevOps-California,Active,4/2/2021,59767.65,bob.davis@example.com,-1651623197,Average,True
1,EMP1001,Bob,Brown,NaN,Finance-Texas,Active,7/10/2020,65304.66,bob.brown@example.com,-1898471390,Excellent,True
2,EMP1002,Alice,Jones,NaN,Admin-Nevada,Pending,12/7/2023,88145.90,alice.jones@example.com,-5596363211,Good,True
3,EMP1003,Eva,Davis,25.0,Admin-Nevada,Inactive,11/27/2021,69450.99,eva.davis@example.com,-3476490784,Good,True
4,EMP1004,Frank,Williams,25.0,Cloud Tech-Florida,Active,1/5/2022,109324.61,frank.williams@example.com,-1586734256,Poor,False


In [108]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1020 entries, 0 to 1019
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Employee_ID        1020 non-null   str    
 1   First_Name         1020 non-null   str    
 2   Last_Name          1020 non-null   str    
 3   Age                809 non-null    float64
 4   Department_Region  1020 non-null   str    
 5   Status             1020 non-null   str    
 6   Join_Date          1020 non-null   str    
 7   Salary             996 non-null    float64
 8   Email              1020 non-null   str    
 9   Phone              1020 non-null   int64  
 10  Performance_Score  1020 non-null   str    
 11  Remote_Work        1020 non-null   bool   
dtypes: bool(1), float64(2), int64(1), str(8)
memory usage: 88.8 KB


In [109]:
df.isna().sum()

Employee_ID            0
First_Name             0
Last_Name              0
Age                  211
Department_Region      0
Status                 0
Join_Date              0
Salary                24
Email                  0
Phone                  0
Performance_Score      0
Remote_Work            0
dtype: int64

## Whitespace and Duplicate Cleaning

In [110]:
df.columns = df.columns.str.strip().str.lower().str.replace(" ", "_")

print(df.columns.tolist())

['employee_id', 'first_name', 'last_name', 'age', 'department_region', 'status', 'join_date', 'salary', 'email', 'phone', 'performance_score', 'remote_work']


In [111]:
string_cols = df.select_dtypes(include=["object", "string"]).columns

for col in string_cols:
    df[col] = df[col].str.strip()

In [142]:
print("Number of duplicated records/rows:", df.duplicated().sum())

Number of duplicated records/rows: 0


## Column Overview, Data Type Conversion, and Format Correction

### `employee_id`

In [113]:
print(df["employee_id"].value_counts())

employee_id
EMP1000    1
EMP1001    1
EMP1002    1
EMP1003    1
EMP1004    1
          ..
EMP2015    1
EMP2016    1
EMP2017    1
EMP2018    1
EMP2019    1
Name: count, Length: 1020, dtype: int64


### `first_name` and `last_name`

In [114]:
print(df["first_name"].unique())
print("")
print(df["last_name"].unique())

<StringArray>
['Bob', 'Alice', 'Eva', 'Frank', 'Charlie', 'David', 'Heidi', 'Grace']
Length: 8, dtype: str

<StringArray>
['Davis', 'Brown', 'Jones', 'Williams', 'Garcia', 'Johnson', 'Miller',
 'Smith']
Length: 8, dtype: str


### `email`

It seems like the email is the combination of first and last name ($8\times 8 = 64$).

In [ ]:
print(df["email"].value_counts())

email
grace.brown@example.com      27
frank.smith@example.com      25
grace.smith@example.com      23
charlie.jones@example.com    22
frank.brown@example.com      22
                             ..
eva.johnson@example.com      10
heidi.davis@example.com      10
heidi.brown@example.com       9
alice.jones@example.com       7
david.jones@example.com       7
Name: count, Length: 64, dtype: int64

In [141]:
print("Number of emails that don't follow the correct email format:", (~df["email"].str.match(r"[\w\.]+@\w+\.\w+")).sum())

Number of emails that don't follow the correct email format: 0


### `age`

In [116]:
print(df["age"].value_counts())

age
40.0    210
25.0    206
30.0    205
35.0    188
Name: count, dtype: int64


In [143]:
print("NaN values inside column 'age':", df["age"].isna().sum())

NaN values inside column 'age': 211


Cleaning the 211 na values requires a bit of EDA to have a reasonable interpolation instead of filling in with median or such constant value. This basic EDA, for example, could be studying for patterns between the employee's name and age.

### `department_region`

In [131]:
print(np.sort(df["department_region"].unique()))

['Admin-California' 'Admin-Florida' 'Admin-Illinois' 'Admin-Nevada'
 'Admin-New York' 'Admin-Texas' 'Cloud Tech-California'
 'Cloud Tech-Florida' 'Cloud Tech-Illinois' 'Cloud Tech-Nevada'
 'Cloud Tech-New York' 'Cloud Tech-Texas' 'DevOps-California'
 'DevOps-Florida' 'DevOps-Illinois' 'DevOps-Nevada' 'DevOps-New York'
 'DevOps-Texas' 'Finance-California' 'Finance-Florida' 'Finance-Illinois'
 'Finance-Nevada' 'Finance-New York' 'Finance-Texas' 'HR-California'
 'HR-Florida' 'HR-Illinois' 'HR-Nevada' 'HR-New York' 'HR-Texas'
 'Sales-California' 'Sales-Florida' 'Sales-Illinois' 'Sales-Nevada'
 'Sales-New York' 'Sales-Texas']


### `status`

In [119]:
print(df["status"].value_counts())

status
Pending     356
Active      352
Inactive    312
Name: count, dtype: int64


### `join_date`

In [120]:
df["join_date"] = pd.to_datetime(df["join_date"], errors="coerce", format="mixed")

In [144]:
print("NaN values inside the formatted column `join_date`:", df["join_date"].isna().sum())

NaN values inside the formatted column `join_date`: 0


In [ ]:
print(df["join_date"].describe())

count                          1020
mean     2022-07-26 12:53:38.823529
min             2020-01-01 00:00:00
25%             2021-04-12 06:00:00
50%             2022-08-09 12:00:00
75%             2023-10-31 12:00:00
max             2024-12-29 00:00:00
Name: join_date, dtype: object

### `salary`

In [ ]:
print(df["salary"].describe())

count       996.000000
mean      85155.056396
std       19873.727918
min       50047.320000
25%       68392.487500
50%       85547.870000
75%      100974.027500
max      119971.650000
Name: salary, dtype: float64

In [123]:
print(df["salary"].isna().sum())

24


Cleaning the 24 na values requires a bit of EDA to have a reasonable interpolation instead of filling in with mean or such constant value. This basic EDA, for example, could be studying for patterns among the employee's department region, age and salary.

### `phone`

In [124]:
df["phone"].describe()

count    1.020000e+03
mean    -4.942253e+09
std      2.817326e+09
min     -9.994973e+09
25%     -7.341992e+09
50%     -4.943997e+09
75%     -2.520391e+09
max     -3.896086e+06
Name: phone, dtype: float64

In [125]:
df["phone"] = df["phone"].abs().astype(str)

In [ ]:
def fill_phone_with_zeros(phone: str):
    """Prepend zeros so that the phone number is a length of 10.
    
    For a given phone number of length 10 or greater, the number itself is returned back.
    """
    length = len(phone)
    zeros_count = 10 - length
    zeros_str = "0" * zeros_count
    return zeros_str + phone

In [127]:
df["phone"] = df["phone"].apply(lambda x: fill_phone_with_zeros(x))
df["phone"] = df["phone"].apply(lambda x: f"{x[:3]}-{x[3:6]}-{x[6:]}")

In [148]:
invalid_phone_mask = ~df["phone"].str.match(r"\d{3}-\d{3}-\d{4}")

print("Count of phone numbers that don't abide the general format:", df.loc[invalid_phone_mask, "phone"].shape[0])

Count of phone numbers that don't abide the general format: 0


### `performance_score`

In [ ]:
print(df["performance_score"].value_counts())

performance_score
Good         270
Average      267
Excellent    267
Poor         216
Name: count, dtype: int64

### `remote_work`

In [134]:
print(df["remote_work"].value_counts())

remote_work
True     513
False    507
Name: count, dtype: int64


## Basic Cleaning EDA